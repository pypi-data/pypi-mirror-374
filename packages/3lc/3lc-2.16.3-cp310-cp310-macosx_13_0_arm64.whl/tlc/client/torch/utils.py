# Pyarmor 9.1.8 (pro), 005329, 3lc, 2025-09-05T11:54:47.851861
from tlcsaas.pyarmor_runtime_005329 import __pyarmor__
__pyarmor__(__name__, __file__, b'PY005329\x00\x03\n\x00o\r\r\n\x80\x00\x01\x00\x08\x00\x00\x00\x04\x00\x00\x00@\x00\x00\x00\x8c\x04\x00\x00\x12\t\x06\x00"\x1dcc\xd3\x95L!\xc5mDG\x0b\x12dW\x00\x00\x00\x00\x00\x00\x00\x00\x17\x05\xd1\xc5\x89\x81D\xddfT\x9f\xf7\xf4v\x19\xa8U\x8a\xa3O\x9e\'\x8d\xff\xc10N\x15\xdf\x03\xa5*\x19sr$\xe7\xc9^\x04^\x85\x98\xb6\xd6\xa8[\xd3\xea~\xff\x08\xb7\x99\xdb\xcc \xf7\x83\x0b\x00\xb5o\xed\xcb\xee\xe6\xedXP"\x1d\xc8\x8a @\x80\xd0\xc0J\xf7g\'\x89\xcc\x8aEM\xda\\\xf3\xb9[\xf8T\r\xc6F\xb4\xfd\x0eS\xa8E\xc7\xb0\xc7>\x16\xa3\xb4\x8eh\xd1\xce\xfb\x80o\xca\x06\x95[wl\x0ex6\xb4_\xb2~\x8e\x8a+\x0c0[\x1an\xe3v\xd1]\xfb"~\xc0\xd2\xf7*4b\x8b&\x04\x1em\xeb\xc3\x0f\xc7z\xe4t\xe4\xc1P\x8a\x0ez\x00c(\x90\xf4~\xbc\x8b\xed1]_\xb0n\xc8\xf3\xaf\x9f\xc4\xb5*\xa4\xe1\xf7\x1b\xe8\x03\xe4|8\x91\xcd\xac\xe5\x1b\xf4\x83\x86,3\xec^\xef}\xb7\xf6\xad\x85\xe45\x8dN2rNT\xb4\x961\xc3\rD\xe0\x1f1P\xd6_*\x97`\xdc\n\xfa)g)S\xa2\x97&\xa6YK3h}\x0e\xa5\xafL\xc7c\xde\xa2\x1d\xbe\xc5/\xa06\xb1)\x96\xca\x82-\xbc\x1f `\xddq\x0eA\xe3\xc9F\x84z\x00N\xea\xe7\x1ad\xfe\xef\\\xae]\xa6\x1bB\x06\x7f\x96\xa6S\xd7\xdf?\xb3\xff\x94|\x91\xcc\xae\x12|p\xd6\xaf\xbc\x10l\x83SD!\xb3n\xa2]$l\xb3\\t\x03\x1dO\x05$>\x1bb\x82\x05b\xa0\x05\xf6f\xc9\xc1\xfe\xd8q\x7f||\xb8\xa6\xde\x82\xe4\xa5n{\x89\xa8\x8a\xa6\xae4B\xf7%\x16y\xcc\x9b\xc2\x80j\xdfS=K\xb1\xc6\xb5\xb1\xb2\x18b\xdfQ\xa2\xe7J\xfe\x9a^3\xc14\xa5u\xd0=\x95\x9893\xa2\x1d\xef\xa4\x08\x17\xd1\x94\x15\x9e)\xee\xf72k\\b\x1bx/q\x00\x95|\xebP\xcd\x9eQ\xfbO4\xaf7\x8f}r\n\x99\x18\xa4hR`\x1b]\x85\x1do\x07\xd44X\x1b\xb9|\xde\xf28\xb1&\xad$.\xb9d\xe1\xdc\x82\x18D9\xd3\xf3-\xdf\x10?G?\xade\x99\x10\x8c\xb3\x890\xa9\xaeMS\x11(\xd7\xec\xf5\x1f\x909\x95,\x1c\xf8W)X\x1fsO\xe4\xef8\xcb\x88\r]\x08q\x0eMy\xd3X\xb5\xe2\x0b*\x94\x87\xf4\x93\xfdr\x14\xd8j\x85\x13\xee)|\xbe\x86\x9a\x0e\xbd5;\xad\xfc\x1c\xbf&\xf2a\xeek;\x8fz \x1e\x04\xea@\xecC\xafg\xe0\x89\x86\xc6 `jc\xbdc_\xa9\x02\xcfXaO\xafF\xe9\x9d\xb4`\x18 \xd8F\xed\xae\x1d1\xadG)\xf8\x97\xc0\xd3\xb2\xc0\xaf\x000\xaa=H8\xbb\x94\xa0_o\x06\\\xa0\xe8\x0b\xb4|\x02\x8f\x8a\xa4\x8c\xf3NE7\x80?\xd8n\xe7\xc6\xf4\x00\x1b\x88\\\x08\x9bw\xa4\xf2\x94kF\x9ed=\x0e\xc1\xfb\xfar\xefr\xb2\x97\xd0\x80o\x91\x1f\xb5h\x99y\x95\xdeD\xbb\xc8\xf2\x9f\x87\x1e\x97\xdb+~\x8f\xc0v\xcb\xd0\xac\xc9\xf2 "a\x07\xb5\xdf\xc4\xba,B\xf2\xa7\xfbP\xf1%&\xd4<H\xbd\xb6\x10F|E\xf9$\xb1\xce\x1d:\xa9\xeb\xe6\x90\xdb\x08S\x01\xddOw\x89\xa6\xcd\xa9\x1eM~\x92\xfcm\x01\x97G\x97h<\xa0\xde?\xfb\xaa\xd0\xc0A\xf8\xbcV=\xe4\xa5\xf7\x12\xa1O\x9a?3\xfdP\\:\x1d\xc2j9\x117G\xe20\xfc\xbc\x93>\xa8\'i\xe8b C\x85Y\xa8\r\x80^\x81\x7f\x1a-\xba\xa4\xa8\xbc\x06\xda\xe6\xacf\xe0\xf6\xbc\x87\xbeND \x8fiJ\x12\\\xbeA\xa2f*\xdbZ\xe4g\x98\xd28\xad\x96\x82\x1eb\xe1\xd9\xd2\xeb%\xfb\xe5\\\xa7\x960F?\xed\x02\x10D\xa1\xf6\xf9\xf3:\xc9\xed\x12x\xdd\xef\xad\xe9\xf7\x81\xab\xf5\x8e\x90\x95\x85\x971o4\x0b\x11\xa7l\xba~(\x87\x8e\xd5\xde\x1d\xee\xdb\xf8\xe7\xf5\xce\x92\xc9\xcd\x07\x80(\xb14<\xe9sX\xd9\xd3sM\x1c\xcda\xdeNV\x83\xdc\xf8\x9b\x99\x99UZ\x9f\x95\xb9I\x02\xben\x9e\x81\xaf\x94\xed[\xdeV\xa9\x07\x04\x89\xdd\xf8\x0c\xa3YY\xe9k\x8azX\xc42\x89\xb2\x8c]In\x11\xf3\xc1\xb2Y\xb5\x9d|MM/6\\G*\xe6`s\xdeb.>\xab\x91e\xf3t\xf5\\g\xecnM \xfe\xfe\xddw\xaf\x83\x0e\x7f\x86\xb4\xa2"\x92U\x11\x85\x85U\x02F\xe0j\x0f/\xbf.d\xcf\x9fsZ\xadp\xeb\'3c_\x91mp\xb3G\xb8J)3\xbcj\xe1Y\xf5\xad\x8e~cH\x85\xf9\x07\xb7\xab\x0b\xcf)\xbc@kO\x18kQ\x99\x8c\xef\xaa\x9ef\xa5\xa6\x0f\x88\xfdk[\xb0]\xa3\xc04\x1b\xf8(\xcb\x99M\x98p"\x88\xc3\xd3\xa4\x08V\xc5\x1b\x16\xebJ\x80\x0b\xac\xc9\n\x7f\xdb\xdd\x1b\xd3?3\xfe\xda8\xf3\x95qVg\xd2\x07\xe6')
