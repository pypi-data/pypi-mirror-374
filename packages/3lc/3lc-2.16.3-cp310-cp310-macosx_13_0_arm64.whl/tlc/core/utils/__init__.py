# Pyarmor 9.1.8 (pro), 005329, 3lc, 2025-09-05T11:54:47.106517
from tlcsaas.pyarmor_runtime_005329 import __pyarmor__
__pyarmor__(__name__, __file__, b'PY005329\x00\x03\n\x00o\r\r\n\x80\x00\x01\x00\x08\x00\x00\x00\x04\x00\x00\x00@\x00\x00\x00\xad\x00\x00\x00\x12\t\x06\x00\xaaO{\x8b\xab\xd8\x18\x97\x880\xc6p\x85\xe3\x88\xcb\x00\x00\x00\x00\x00\x00\x00\x00\x98\xd7\xafR\xe1\xc8F$\xa1s\xfe \xa1N\x03B\x0fe\x8b\x964\x8704Z\xb6:\xfa 5\xe2Of\xa2^\xf0\xde\x90\x86\xf0\x1ch\t\xaa\x91\xde\xcc\\\xc2& \xe2%\xd5\x90&\x1b(OI/\x1e\xc3H!\xd3\x1d\r\xd74B\x02D\x87e\xd2\xbc\xce.\xb7\xce]5\xf6BPp\x91\x18\xc4\xa69\xe5\x94\x16\xd6\xadlv\x8c\xa5\xedV>\x08.\x11\xd5\xffDe\x14Tp\xfc\x93\xab=\xd8{\xa7L)\xc3\x1dN_\x9dZ\x9c\xcb\xc8\xa30|\xc9\xf8\xd3\xd4l\xa1\x86j\xf9\xda7\x95\x1f\xe1\tG\x82-\xc8[5\x1fs\xc06\xe8\x11\xd8\x18\x83R5s\xac\nOr\x9d')
