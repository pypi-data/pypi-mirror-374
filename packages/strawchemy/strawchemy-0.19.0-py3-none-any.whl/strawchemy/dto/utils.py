"""Utilities for DTO configuration and type checking.

This module provides utility functions for configuring Data Transfer Objects (DTOs)
and performing type checks, specifically for optional type hints.

It exports the following:

- `config`: Configures a DTOConfig object for a specific purpose.
- `field`: Configures a DTOFieldConfig object for a specific purpose.
- `is_type_hint_optional`: Checks if a type hint is optional.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional

from .constants import DTO_INFO_KEY
from .types import DTOConfig, DTOFieldConfig, DTOScope, ExcludeFields, IncludeFields, Purpose, PurposeConfig

if TYPE_CHECKING:
    from collections.abc import Callable, Mapping

__all__ = (
    "PRIVATE",
    "READ_ONLY",
    "WRITE_ONLY",
    "config",
    "field",
    "read_all_config",
    "read_all_partial_config",
    "read_partial",
    "write_all_config",
    "write_all_partial_config",
)


def config(
    purpose: Purpose,
    include: Optional[IncludeFields] = None,
    exclude: Optional[ExcludeFields] = None,
    partial: Optional[bool] = None,
    type_map: Optional[Mapping[Any, Any]] = None,
    aliases: Optional[Mapping[str, str]] = None,
    alias_generator: Optional[Callable[[str], str]] = None,
    scope: Optional[DTOScope] = None,
    tags: Optional[set[str]] = None,
) -> DTOConfig:
    config = DTOConfig(purpose, alias_generator=alias_generator, scope=scope)
    if exclude:
        config.exclude = exclude
    if include:
        config.include = include
    if type_map:
        config.type_overrides = type_map
    if aliases:
        config.aliases = aliases
    if partial is not None:
        config.partial = partial
    if tags:
        config.tags = tags
    return config


def field(
    purposes: Optional[set[Purpose]] = None,
    default_config: Optional[PurposeConfig] = None,
    configs: Optional[dict[Purpose, PurposeConfig]] = None,
) -> dict[str, DTOFieldConfig]:
    return {
        DTO_INFO_KEY: DTOFieldConfig(
            purposes=purposes if purposes is not None else {Purpose.READ, Purpose.WRITE},
            default_config=default_config or PurposeConfig(),
            configs=configs or {},
        ),
    }


read_partial = DTOConfig(Purpose.READ, partial=True)
read_all_config = DTOConfig(Purpose.READ, include="all")
read_all_partial_config = DTOConfig(Purpose.READ, include="all", partial=True)
write_all_config = DTOConfig(Purpose.WRITE, include="all")
write_all_partial_config = DTOConfig(Purpose.WRITE, include="all", partial=True)

READ_ONLY = field({Purpose.READ})
WRITE_ONLY = field({Purpose.WRITE})
PRIVATE = field(set())
