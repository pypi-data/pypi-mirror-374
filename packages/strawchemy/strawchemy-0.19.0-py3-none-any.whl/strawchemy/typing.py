from __future__ import annotations

import sys
from typing import TYPE_CHECKING, Any, ClassVar, Literal, Protocol, Union

from typing_extensions import TypeAlias

if sys.version_info >= (3, 10):
    from types import UnionType

    UNION_TYPES = (Union, UnionType)
else:
    UNION_TYPES = (Union,)


if TYPE_CHECKING:
    from . import StrawchemyAsyncRepository, StrawchemySyncRepository

__all__ = ("UNION_TYPES", "AnyRepository", "DataclassProtocol", "SupportedDialect")


class DataclassProtocol(Protocol):
    __dataclass_fields__: ClassVar[dict[str, Any]]


AnyRepository: TypeAlias = "type[StrawchemySyncRepository[Any] | StrawchemyAsyncRepository[Any]]"
SupportedDialect: TypeAlias = Literal["postgresql", "mysql", "sqlite"]
"""Must match SQLAlchemy dialect."""
