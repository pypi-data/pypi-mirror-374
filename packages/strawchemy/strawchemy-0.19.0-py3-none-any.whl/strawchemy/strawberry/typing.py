from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, TypeVar

from typing_extensions import TypeAlias

if TYPE_CHECKING:
    from collections.abc import Callable

    from sqlalchemy import Select
    from strawberry import Info
    from strawberry.types.base import WithStrawberryObjectDefinition
    from strawchemy.graph import Node
    from strawchemy.sqlalchemy.typing import AnyAsyncSession, AnySyncSession
    from strawchemy.strawberry.dto import GraphQLFieldDefinition, QueryNodeMetadata, StrawchemyDTOAttributes
    from strawchemy.validation.pydantic import MappedPydanticGraphQLDTO

    from .dto import (
        AggregateDTO,
        FilterFunctionInfo,
        GraphQLFilterDTO,
        MappedStrawberryGraphQLDTO,
        OrderByDTO,
        OutputFunctionInfo,
        UnmappedStrawberryGraphQLDTO,
    )

__all__ = (
    "AnySessionGetter",
    "AsyncSessionGetter",
    "FilterStatementCallable",
    "StrawchemyTypeWithStrawberryObjectDefinition",
    "SyncSessionGetter",
)


_T = TypeVar("_T")
QueryObject = TypeVar("QueryObject", bound="Any")
GraphQLFilterDTOT = TypeVar("GraphQLFilterDTOT", bound="GraphQLFilterDTO")
AggregateDTOT = TypeVar("AggregateDTOT", bound="AggregateDTO")
GraphQLDTOT = TypeVar("GraphQLDTOT", bound="GraphQLDTO[Any]")
OrderByDTOT = TypeVar("OrderByDTOT", bound="OrderByDTO")

AggregationFunction = Literal["min", "max", "sum", "avg", "count", "stddev_samp", "stddev_pop", "var_samp", "var_pop"]
AggregationType = Literal[
    "sum", "numeric", "min_max_datetime", "min_max_date", "min_max_time", "min_max_string", "min_max_numeric"
]

GraphQLType = Literal["input", "object", "interface", "enum"]
AsyncSessionGetter: TypeAlias = "Callable[[Info[Any, Any]], AnyAsyncSession]"
SyncSessionGetter: TypeAlias = "Callable[[Info[Any, Any]], AnySyncSession]"
AnySessionGetter: TypeAlias = "AsyncSessionGetter | SyncSessionGetter"
FilterStatementCallable: TypeAlias = "Callable[[Info[Any, Any]], Select[tuple[Any]]]"
GraphQLPurpose: TypeAlias = Literal[
    "type",
    "aggregate_type",
    "create_input",
    "update_by_pk_input",
    "update_by_filter_input",
    "filter",
    "aggregate_filter",
    "order_by",
    "upsert_update_fields",
    "upsert_conflict_fields",
]
FunctionInfo: TypeAlias = "FilterFunctionInfo | OutputFunctionInfo"
StrawberryGraphQLDTO: TypeAlias = "MappedStrawberryGraphQLDTO[_T] | UnmappedStrawberryGraphQLDTO[_T]"
GraphQLDTO: TypeAlias = "StrawberryGraphQLDTO[_T] | MappedPydanticGraphQLDTO[_T]"
MappedGraphQLDTO: TypeAlias = "MappedStrawberryGraphQLDTO[_T] | MappedPydanticGraphQLDTO[_T]"
AnyMappedDTO: TypeAlias = "MappedStrawberryGraphQLDTO[Any] | MappedPydanticGraphQLDTO[Any]"
QueryNodeType: TypeAlias = "Node[GraphQLFieldDefinition, QueryNodeMetadata]"

if TYPE_CHECKING:

    class StrawchemyTypeWithStrawberryObjectDefinition(StrawchemyDTOAttributes, WithStrawberryObjectDefinition): ...
