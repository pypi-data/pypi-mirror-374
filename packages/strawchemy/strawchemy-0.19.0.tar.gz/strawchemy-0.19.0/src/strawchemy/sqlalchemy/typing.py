from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, TypeVar

from typing_extensions import TypeAlias

if TYPE_CHECKING:
    from collections import OrderedDict
    from collections.abc import Callable

    from sqlalchemy import Function
    from sqlalchemy.ext.asyncio import AsyncSession, async_scoped_session
    from sqlalchemy.orm import DeclarativeBase, Session, scoped_session
    from sqlalchemy.sql import SQLColumnExpression
    from strawchemy.strawberry.dto import OrderByEnum
    from strawchemy.strawberry.filters.base import GraphQLComparison

    from ._executor import QueryExecutor
    from .hook import QueryHook


__all__ = (
    "AnyAsyncSession",
    "AnySession",
    "AnySyncSession",
    "DeclarativeSubT",
    "DeclarativeT",
    "FunctionGenerator",
    "QueryExecutorT",
    "QueryHookCallable",
    "RelationshipSide",
    "SessionT",
    "StatementType",
)

DeclarativeT = TypeVar("DeclarativeT", bound="DeclarativeBase")
DeclarativeSubT = TypeVar("DeclarativeSubT", bound="DeclarativeBase")
QueryHookDeclarativeT = TypeVar("QueryHookDeclarativeT", bound="DeclarativeBase")
SessionT = TypeVar("SessionT", bound="AnySession")
QueryExecutorT = TypeVar("QueryExecutorT", bound="QueryExecutor[Any]")

RelationshipSide: TypeAlias = Literal["parent", "target"]
StatementType = Literal["lambda", "select"]
FunctionGenerator: TypeAlias = "Callable[..., Function[Any]]"
QueryHookCallable: TypeAlias = "QueryHook[QueryHookDeclarativeT]"
FilterMap: TypeAlias = "OrderedDict[tuple[type[Any], ...], type[GraphQLComparison]]"
AnySyncSession: TypeAlias = "Session | scoped_session[Session]"
AnyAsyncSession: TypeAlias = "AsyncSession | async_scoped_session[AsyncSession]"
AnySession: TypeAlias = "AnySyncSession | AnyAsyncSession"
OrderBySpec: TypeAlias = "tuple[SQLColumnExpression[Any], OrderByEnum]"
