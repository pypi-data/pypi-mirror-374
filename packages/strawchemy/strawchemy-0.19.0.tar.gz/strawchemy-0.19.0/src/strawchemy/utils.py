from __future__ import annotations

import inspect
import re
import sys
from typing import TYPE_CHECKING, Any, Optional, Union, get_args, get_origin

from strawchemy.typing import UNION_TYPES

if TYPE_CHECKING:
    from re import Pattern

__all__ = (
    "camel_to_snake",
    "is_type_hint_optional",
    "non_optional_type_hint",
    "snake_keys",
    "snake_to_camel",
    "snake_to_lower_camel_case",
)

_camel_to_snake_pattern: Pattern[str] = re.compile(r"((?<=[a-z0-9])[A-Z]|(?!^)(?<!_)[A-Z](?=[a-z]))")


def camel_to_snake(string: str) -> str:
    """Convert a camelcased string to snake case.

    See: https://stackoverflow.com/a/12867228
    """
    return _camel_to_snake_pattern.sub(r"_\1", string).lower()


def snake_to_camel(string: str) -> str:
    """Convert string to camel case.

    See: https://stackoverflow.com/a/19053800/10735573
    """
    return "".join(x.capitalize() for x in string.lower().split("_"))


def snake_to_lower_camel_case(snake_str: str) -> Any:
    """Convert string to lower camel case.

    See: https://stackoverflow.com/a/19053800/10735573
    """
    camel_string: str = snake_to_camel(snake_str)
    return snake_str[0].lower() + camel_string[1:]


def snake_keys(value: dict[str, Any]) -> dict[str, Any]:
    """Recursively convert dict keys to from camel case to snake case."""
    res: dict[Any, Any] = {}
    for k, v in value.items():
        to_snake: str = camel_to_snake(k)
        if isinstance(v, (list, tuple)):
            res[to_snake] = [snake_keys(el) for el in v]
        elif isinstance(v, dict):
            res[to_snake] = snake_keys(v)
        else:
            res[to_snake] = v
    return res


def non_optional_type_hint(type_hint: Any) -> Any:
    origin, args = get_origin(type_hint), get_args(type_hint)
    if origin is Optional:
        return args
    if origin in UNION_TYPES:
        union_args = tuple([arg for arg in args if arg not in (None, type(None))])
        if len(union_args) == 1:
            return union_args[0]
        return Union[union_args]
    return type_hint


def is_type_hint_optional(type_hint: Any) -> bool:
    """Whether the given type hint is considered as optional or not.

    Returns:
        `True` if arguments of the given type hint are optional

    Three cases are considered:
    ```
        Optional[str]
        Union[str, None]
        str | None
    ```
    In any other form, the type hint will not be considered as optional
    """
    origin = get_origin(type_hint)
    if origin is None:
        return False
    if origin is Optional:
        return True
    if origin in UNION_TYPES:
        args = get_args(type_hint)
        return any(arg is type(None) for arg in args)
    return False


def get_annotations(obj: Any) -> dict[str, Any]:
    """Get the annotations of the given object."""
    if sys.version_info >= (3, 10):
        return inspect.get_annotations(obj)
    return obj.__dict__.get("__annotations__", {}) if isinstance(obj, type) else getattr(obj, "__annotations__", {})
