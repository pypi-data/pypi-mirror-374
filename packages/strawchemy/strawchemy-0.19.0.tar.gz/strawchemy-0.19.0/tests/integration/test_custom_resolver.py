from __future__ import annotations

from typing import TYPE_CHECKING, Literal

import pytest

from tests.integration.utils import to_graphql_representation
from tests.typing import AnyQueryExecutor
from tests.utils import maybe_async

if TYPE_CHECKING:
    from syrupy.assertion import SnapshotAssertion

    from .fixtures import QueryTracker


pytestmark = [pytest.mark.integration]


@pytest.mark.snapshot
async def test_get_one(
    any_query: AnyQueryExecutor, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    result = await maybe_async(any_query("{ redColor { name } }"))

    assert not result.errors
    assert result.data
    assert result.data["redColor"] == {"name": "Red"}

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.parametrize("color", ["unknown", "Pink"])
@pytest.mark.snapshot
async def test_get_one_or_none(color: Literal["unknown", "Pink"], any_query: AnyQueryExecutor) -> None:
    result = await maybe_async(
        any_query(f"""
            {{
                  getColor(color: {to_graphql_representation(color, "input")}) {{
                    name
                }}
            }}
        """)
    )

    assert not result.errors
    assert result.data
    assert result.data["getColor"] == ({"name": "Pink"} if color == "Pink" else None)
