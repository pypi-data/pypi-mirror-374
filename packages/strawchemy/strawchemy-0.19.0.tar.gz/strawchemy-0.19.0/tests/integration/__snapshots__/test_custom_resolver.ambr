# serializer version: 1
# name: test_get_one[session-tracked-async-aiosqlite_engine]
  '''
  SELECT color.name,
         color.id
    FROM color AS color
    JOIN (
          SELECT color.name AS name,
                 color.created_at AS created_at,
                 color.updated_at AS updated_at,
                 color.id AS id
            FROM color
           WHERE color.name = ?
         ) AS anon_1
      ON color.id = anon_1.id
  '''
# ---
# name: test_get_one[session-tracked-async-asyncmy_engine]
  '''
  SELECT color.name,
         color.id
    FROM color AS color
   INNER JOIN (
          SELECT color.name AS name,
                 color.created_at AS created_at,
                 color.updated_at AS updated_at,
                 color.id AS id
            FROM color
           WHERE color.name = %s
         ) AS anon_1
      ON color.id = anon_1.id
  '''
# ---
# name: test_get_one[session-tracked-async-asyncpg_engine]
  '''
  SELECT color.name,
         color.id
    FROM color AS color
    JOIN (
          SELECT color.name AS name,
                 color.created_at AS created_at,
                 color.updated_at AS updated_at,
                 color.id AS id
            FROM color
           WHERE color.name = $1::VARCHAR
         ) AS anon_1
      ON color.id = anon_1.id
  '''
# ---
# name: test_get_one[session-tracked-async-psycopg_async_engine]
  '''
  SELECT color.name,
         color.id
    FROM color AS color
    JOIN (
          SELECT color.name AS name,
                 color.created_at AS created_at,
                 color.updated_at AS updated_at,
                 color.id AS id
            FROM color
           WHERE color.name = %(name_1)s::VARCHAR
         ) AS anon_1
      ON color.id = anon_1.id
  '''
# ---
# name: test_get_one[session-tracked-sync-psycopg_engine]
  '''
  SELECT color.name,
         color.id
    FROM color AS color
    JOIN (
          SELECT color.name AS name,
                 color.created_at AS created_at,
                 color.updated_at AS updated_at,
                 color.id AS id
            FROM color
           WHERE color.name = %(name_1)s::VARCHAR
         ) AS anon_1
      ON color.id = anon_1.id
  '''
# ---
# name: test_get_one[session-tracked-sync-sqlite_engine]
  '''
  SELECT color.name,
         color.id
    FROM color AS color
    JOIN (
          SELECT color.name AS name,
                 color.created_at AS created_at,
                 color.updated_at AS updated_at,
                 color.id AS id
            FROM color
           WHERE color.name = ?
         ) AS anon_1
      ON color.id = anon_1.id
  '''
# ---
